src/lib/math/pcurves/pcurves_impl.h

```diff
@@ -0,0 +1,1469 @@
+            } else {
+               table.push_back(table[i - 1] + table[0]);
+            }
+         }
+
+         m_table = ProjectivePoint::to_affine_batch(table);
+      }
+
+      ProjectivePoint mul(const Scalar& s, RandomNumberGenerator& rng) const {
+         const BlindedScalar bits(s, rng);
+
+         auto accum = [&]() {
+            const size_t w_0 = bits.get_window((Windows - 1) * WindowBits);
+            // Guaranteed because we set the high bit of the randomizer
+            BOTAN_DEBUG_ASSERT(w_0 != 0);
+            auto pt = ProjectivePoint::from_affine(AffinePoint::ct_select(m_table, w_0));
+            pt.ct_poison();
+            pt.randomize_rep(rng);
+            return pt;
+         }();
+
+         for(size_t i = 1; i != Windows; ++i) {
+            accum = accum.dbl_n(WindowBits);
+            const size_t w_i = bits.get_window((Windows - i - 1) * WindowBits);
+
+            /*
+            This point addition cannot be a doubling (except once)
+
+            Consider the sequence of points that are operated on, and specifically
+            their discrete logarithms. We start out at the point at infinity
+            (dlog 0) and then add the initial window which is precisely P*w_0
+
+            We then perform WindowBits doublings, so accum's dlog at the point
+            of the addition in the first iteration of the loop (when i == 1) is
+            at least 2^W * w_0.
+
+            Since we know w_0 > 0, then in every iteration of the loop, accums
+            dlog will always be greater than the dlog of the table element we
+            just looked up (something between 0 and 2^W-1), and thus the
+            addition into accum cannot be a doubling.
+
+            However due to blinding this argument fails, since we perform
+            multiplications using a scalar that is larger than the group
+            order. In this case it's possible that the dlog of accum becomes
+            `order + x` (or, effectively, `x`) and `x` is smaller than 2^W.
+            In this case, a doubling may occur. Future iterations of the loop
+            cannot be doublings by the same argument above. Since the blinding
+            factor is always less than the group order (substantially so),
+            it is not possible for the dlog of accum to overflow a second time.
+            */
+            accum += AffinePoint::ct_select(m_table, w_i);
+
+            if(i <= 3) {
+               accum.randomize_rep(rng);
+            }
+         }
+
+         accum.ct_unpoison();
+         return accum;
+      }
+
+   private:
+      std::vector<AffinePoint> m_table;
+};
+
+/**
+* Effect 2-ary multiplication ie x*G + y*H
+*
+* This is done using a windowed variant of what is usually called
+* Shamir's trick.
+*
+* The W = 1 case is simple; we precompute an extra point GH = G + H,
+* and then examine 1 bit in each of x and y. If one or the other bits
+* are set then add G or H resp. If both bits are set, add GH.
```

Comment:
In #4095 we added a bit of ASCII art for the W = 2 case. Perhaps that's worth adding:

```C++
     /**
       * The example below is a precomputed table for W=2. The flattened table
       * begins at (x_i,y_i) = (1,0), i.e. the identity element is omitted.
       * The indices in each cell refer to the cell's location in m_table.
       *
       *  x->           0          1          2         3
       *       0  |/ (ident) |0  x     |1  2x      |2  3x     |
       *       1  |3    y    |4  x+y   |5  2x+y    |6  3x+y   |
       *  y =  2  |7    2y   |8  x+2y  |9  2(x+y)  |10 3x+2y  |
       *       3  |11   3y   |12 x+3y  |13 2x+3y   |14 3x+3y  |
       */
```

Reply:
(this is also part of https://github.com/randombit/botan/pull/3979#discussion_r1630787145)

Reply:
Added this comment